integration tests flake due to minio leader not being unit 0

### Bug Description

When the integration tests interact with a unit of a deployed application they use the hardcoded unit 0. This typically works, but sometimes applications dont have a unit 0 (maybe because 0 was killed and 1 was created), causing test failures as seen [here](https://github.com/canonical/mimir-coordinator-k8s-operator/actions/runs/10475134244/job/29011121957?pr=68#step:7:897) where we see the message:
```
  File "/home/runner/work/mimir-coordinator-k8s-operator/mimir-coordinator-k8s-operator/tests/integration/helpers.py", line 26, in configure_minio
    minio_addr = await get_unit_address(ops_test, "minio", 0)
  File "/home/runner/work/mimir-coordinator-k8s-operator/mimir-coordinator-k8s-operator/tests/integration/helpers.py", line 68, in get_unit_address
    assert False, f"no unit exists in app {app_name} with index {unit_no}"
AssertionError: no unit exists in app minio with index 0
```

### To Reproduce

-

### Environment

-

### Relevant log output

```shell
File "/home/runner/work/mimir-coordinator-k8s-operator/mimir-coordinator-k8s-operator/tests/integration/helpers.py", line 26, in configure_minio
    minio_addr = await get_unit_address(ops_test, "minio", 0)
  File "/home/runner/work/mimir-coordinator-k8s-operator/mimir-coordinator-k8s-operator/tests/integration/helpers.py", line 68, in get_unit_address
    assert False, f"no unit exists in app {app_name} with index {unit_no}"
AssertionError: no unit exists in app minio with index 0
```


### Additional context

_No response_

--- a/tests/integration/helpers.py
+++ b/tests/integration/helpers.py
@@ -23,7 +23,8 @@ def charm_resources(metadata_file="metadata.yaml") -> Dict[str, str]:
 
 async def configure_minio(ops_test: OpsTest):
     bucket_name = "mimir"
-    minio_addr = await get_unit_address(ops_test, "minio", 0)
+    minio_leader_unit_number = await get_leader_unit_number(ops_test, "minio")
+    minio_addr = await get_unit_address(ops_test, "minio", minio_leader_unit_number)
     mc_client = Minio(
         f"{minio_addr}:9000",
         access_key="access",
@@ -57,6 +58,23 @@ async def configure_s3_integrator(ops_test: OpsTest):
     assert action_result.status == "completed"
 
 
+async def get_leader_unit_number(ops_test: OpsTest, app_name: str) -> int:
+    """Get the unit number of the leader of an application.
+
+    Raises an exception if no leader is found.
+    """
+    status = await ops_test.model.get_status()
+    app = status["applications"][app_name]
+    if app is None:
+        raise ValueError(f"no app exists with name {app_name}")
+
+    for name, unit in app["units"].items():
+        if unit["leader"]:
+            return int(name.split("/")[1])
+
+    raise ValueError(f"no leader found for app {app_name}")
+
+
 async def get_unit_address(ops_test: OpsTest, app_name: str, unit_no: int) -> str:
     assert ops_test.model is not None
     status = await ops_test.model.get_status()
@@ -79,7 +97,8 @@ async def get_grafana_datasources(ops_test: OpsTest, grafana_app: str = "grafana
     action = await grafana_leader.run_action("get-admin-password")
     action_result = await action.wait()
     admin_password = action_result.results["admin-password"]
-    grafana_url = await get_unit_address(ops_test, grafana_app, 0)
+    leader_unit_number = await get_leader_unit_number(ops_test, grafana_app)
+    grafana_url = await get_unit_address(ops_test, grafana_app, leader_unit_number)
     response = requests.get(f"http://admin:{admin_password}@{grafana_url}:3000/api/datasources")
     assert response.status_code == 200
 
@@ -95,7 +114,8 @@ async def get_prometheus_targets(
         {"status": "success", "data": {"activeTargets": [{"discoveredLabels": {..., "juju_charm": <charm>, ...}}]}}
     """
     assert ops_test.model is not None
-    prometheus_url = await get_unit_address(ops_test, prometheus_app, 0)
+    leader_unit_number = await get_leader_unit_number(ops_test, prometheus_app)
+    prometheus_url = await get_unit_address(ops_test, prometheus_app, leader_unit_number)
     response = requests.get(f"http://{prometheus_url}:9090/api/v1/targets")
     assert response.status_code == 200
     assert response.json()["status"] == "success"
@@ -106,7 +126,8 @@ async def get_prometheus_targets(
 async def query_mimir(
     ops_test: OpsTest, query: str, coordinator_app: str = "mimir"
 ) -> Dict[str, Any]:
-    mimir_url = await get_unit_address(ops_test, coordinator_app, 0)
+    leader_unit_number = await get_leader_unit_number(ops_test, coordinator_app)
+    mimir_url = await get_unit_address(ops_test, coordinator_app, leader_unit_number)
     response = requests.get(
         f"http://{mimir_url}:8080/prometheus/api/v1/query",
         params={"query": query},

--- a/tests/integration/test_charm_ha.py
+++ b/tests/integration/test_charm_ha.py
@@ -32,11 +32,11 @@ async def test_build_and_deploy(ops_test: OpsTest, mimir_charm: str):
     assert ops_test.model is not None  # for pyright
     await asyncio.gather(
         ops_test.model.deploy(mimir_charm, "mimir", resources=charm_resources()),
-        ops_test.model.deploy("prometheus-k8s", "prometheus", channel="latest/edge"),
-        ops_test.model.deploy("loki-k8s", "loki", channel="latest/edge"),
-        ops_test.model.deploy("grafana-k8s", "grafana", channel="latest/edge"),
+        ops_test.model.deploy("prometheus-k8s", "prometheus", channel="latest/edge", trust=True),
+        ops_test.model.deploy("loki-k8s", "loki", channel="latest/edge", trust=True),
+        ops_test.model.deploy("grafana-k8s", "grafana", channel="latest/edge", trust=True),
         ops_test.model.deploy("grafana-agent-k8s", "agent", channel="latest/edge"),
-        ops_test.model.deploy("traefik-k8s", "traefik", channel="latest/edge"),
+        ops_test.model.deploy("traefik-k8s", "traefik", channel="latest/edge", trust=True),
         # Deploy and configure Minio and S3
         # Secret must be at least 8 characters: https://github.com/canonical/minio-operator/issues/137
         ops_test.model.deploy(

--- a/tests/integration/test_charm_ha_scaled.py
+++ b/tests/integration/test_charm_ha_scaled.py
@@ -31,11 +31,11 @@ async def test_build_and_deploy(ops_test: OpsTest, mimir_charm: str):
     assert ops_test.model is not None  # for pyright
     await asyncio.gather(
         ops_test.model.deploy(mimir_charm, "mimir", resources=charm_resources()),
-        ops_test.model.deploy("prometheus-k8s", "prometheus", channel="latest/edge"),
-        ops_test.model.deploy("loki-k8s", "loki", channel="latest/edge"),
-        ops_test.model.deploy("grafana-k8s", "grafana", channel="latest/edge"),
+        ops_test.model.deploy("prometheus-k8s", "prometheus", channel="latest/edge", trust=True),
+        ops_test.model.deploy("loki-k8s", "loki", channel="latest/edge", trust=True),
+        ops_test.model.deploy("grafana-k8s", "grafana", channel="latest/edge", trust=True),
         ops_test.model.deploy("grafana-agent-k8s", "agent", channel="latest/edge"),
-        ops_test.model.deploy("traefik-k8s", "traefik", channel="latest/edge"),
+        ops_test.model.deploy("traefik-k8s", "traefik", channel="latest/edge", trust=True),
         # Deploy and configure Minio and S3
         # Secret must be at least 8 characters: https://github.com/canonical/minio-operator/issues/137
         ops_test.model.deploy(

--- a/tests/integration/test_charm_monolithic.py
+++ b/tests/integration/test_charm_monolithic.py
@@ -31,11 +31,11 @@ async def test_build_and_deploy(ops_test: OpsTest, mimir_charm: str):
     assert ops_test.model is not None  # for pyright
     await asyncio.gather(
         ops_test.model.deploy(mimir_charm, "mimir", resources=charm_resources()),
-        ops_test.model.deploy("prometheus-k8s", "prometheus", channel="latest/edge"),
-        ops_test.model.deploy("loki-k8s", "loki", channel="latest/edge"),
-        ops_test.model.deploy("grafana-k8s", "grafana", channel="latest/edge"),
+        ops_test.model.deploy("prometheus-k8s", "prometheus", channel="latest/edge", trust=True),
+        ops_test.model.deploy("loki-k8s", "loki", channel="latest/edge", trust=True),
+        ops_test.model.deploy("grafana-k8s", "grafana", channel="latest/edge", trust=True),
         ops_test.model.deploy("grafana-agent-k8s", "agent", channel="latest/edge"),
-        ops_test.model.deploy("traefik-k8s", "traefik", channel="latest/edge"),
+        ops_test.model.deploy("traefik-k8s", "traefik", channel="latest/edge", trust=True),
         # Deploy and configure Minio and S3
         # Secret must be at least 8 characters: https://github.com/canonical/minio-operator/issues/137
         ops_test.model.deploy(

